table/tables/mutation_checker.go

```diff
@@ -0,0 +1,244 @@
+func CheckIndexConsistency(sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded, dataRemoved []types.Datum, memBuffer kv.MemBuffer, sh kv.StagingHandle) error {
+	sc := sessVars.StmtCtx
+	if sh == 0 {
+		return nil
+	}
+	mutations := collectTableMutationsFromBufferStage(t, memBuffer, sh)
+	if err := checkRowAdditionConsistency(sessVars, t.Meta().Columns, dataAdded, mutations); err != nil {
+		return errors.Trace(err)
+	}
+	if err := checkIndexKeys(sc, sessVars, t, dataAdded, dataRemoved, mutations); err != nil {
+		return errors.Trace(err)
+	}
+	// TODO: check whether handles match in index and row mutations
+	return nil
+}
+
+// checkIndexKeys checks whether the decoded data from keys of index mutations are consistent with the expected ones.
+func checkIndexKeys(sc *stmtctx.StatementContext, sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded []types.Datum, dataRemoved []types.Datum, mutations []mutation) error {
+	indexIDMap := make(map[int64]indexHelperInfo)
+	for _, index := range t.indices {
+		indexIDMap[index.Meta().ID] = indexHelperInfo{
+			index.Meta(),
+			BuildRowcodecColInfoForIndexColumns(index.Meta(), t.Meta()),
+		}
+	}
+
+	for _, m := range mutations {
+		if !tablecodec.IsIndexKey(m.key) {
+			continue
+		}
+
+		_, indexID, _, err := tablecodec.DecodeIndexKey(m.key)
+		if err != nil {
+			continue
+		}
+
+		indexHelperInfo, ok := indexIDMap[indexID]
+		if !ok {
+			return errors.New("index not found")
+		}
+
+		// when we cannot decode the key to get the original value
+		if len(m.value) == 0 && NeedRestoredData(indexHelperInfo.indexInfo.Columns, t.Meta().Columns) {
+			continue
+		}
+
+		decodedIndexValues, err := tablecodec.DecodeIndexKV(m.key, m.value, len(indexHelperInfo.indexInfo.Columns),
+			tablecodec.HandleNotNeeded, indexHelperInfo.rowColInfos)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		indexData := make([]types.Datum, 0)
+		for i, v := range decodedIndexValues {
+			fieldType := &t.Columns[indexHelperInfo.indexInfo.Columns[i].Offset].FieldType
+			datum, err := tablecodec.DecodeColumnValue(v, fieldType, sessVars.Location())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			indexData = append(indexData, datum)
+		}
+
+		if len(m.value) == 0 {
+			err = compareIndexData(sc, t.Columns, indexData, dataRemoved, indexHelperInfo.indexInfo)
+		} else {
+			err = compareIndexData(sc, t.Columns, indexData, dataAdded, indexHelperInfo.indexInfo)
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// checkRowAdditionConsistency checks whether the values of row mutations are consistent with the expected ones
+// We only check data added since a deletion of a row doesn't care about its value (and we cannot know it)
+func checkRowAdditionConsistency(sessVars *variable.SessionVars, tableColumns []*model.ColumnInfo, dataAdded []types.Datum, mutations []mutation) error {
+	rowsAdded, _ := extractRowMutations(mutations)
+	if dataAdded == nil {
+		// should be unreachable
```

Comment:
For the early stage we could print `ERROR` log and return an error here.

Reply:
The comment is wrong. It's possible when the operation is a deletion. I've fixed it